[ARM] Fix llvm.returnaddress for Thumb1 with R11 frame-pointer

ostannard opened the pull request:
When using R11 as the frame pointer for a Thumb1 target, we need to use LR as a temporary in the prologue when setting up the frame record, like this:
```
  push {lr}
  mov lr, r11
  push {r11}
  mov r11, sp
```

This means that the value in LR is clobbered by the prologue, so we can't use it as a live-in value to the function to implement the llvm.returnaddress intrinsic. I don't think there's any way to avoid this use if LR without making the prologue code even more complicated. Instead, we can load the return address from the frame record, as we do when the intrinsic is used with a non-zero depth.

efriedma-quic commented:
I'm concerned that this is fragile: isel is guessing what framelowering is going to do, and if it guesses wrong, you get a miscompile.  And there's no obvious connection between the bits of code in question.

Either framelowering should preserve LR by using a different code sequence if isReturnAddressTaken() is set, or the decision on how to lower returnaddress should be deferred to framelowering (using a pseudo-instruction).

ostannard committed:
Test showing incorrect codegen

ostannard committed:
[ARM] Fix llvm.returnaddress for Thumb1 with R11 frame-pointer

When the llvm.returnaddress intrinsic is used, the LR is marked as
live-in to the function, so it must be preserved through the prologue.
This is normally fine, but there is one case for Thumb1 where we use LR
as a temporary in the prologue to set up a frame chain using r11 as the
frame pointer. There are no other registers guaranteed to be free to do
this, so we have to re-load LR from the stack after pushing the callee
saved registers.

ostannard committed:
[ARM] Optimise Thumb1 frame record creation to not clobber LR

When setting up frame records using r11 as the frame pointer for Thumb1,
we currently use LR as a temporary. This requires us to re-load it in
cases where LR is live into the function, so it is better to use an
argument register if one is free.

ostannard committed:
clang-format

efriedma-quic reviewed (approved):
For the pseudo, I was thinking it would be sort of opportunistic: if the return address happens to be available in LR, then mov from LR, else load it from the stack.  (If it's not in LR, it's almost always going to be somewhere on the stack.)  I guess that's potentially less efficient in some cases, but probably fine for Thumb1.

In any case, LGTM.
